File: skills/plotting/python/pavlab_stripchart.py
"""pavlab_stripchart — opinionated lab-style strip chart.

One categorical axis, one continuous axis, points jittered horizontally.
Shares all style helpers with pavlab_scatter.

Usage
-----
    from pavlab_stripchart import pavlab_stripchart

    # basic: all black points
    pavlab_stripchart(groups, values, xlabel="Condition", ylabel="Expression")

    # categorical colour → lab accent palette + legend
    pavlab_stripchart(groups, values, color=cell_type_labels)

    # numeric colour → viridis colorbar
    pavlab_stripchart(groups, values, color=scores, color_label="score")

    # explicit group order + grand-mean line
    pavlab_stripchart(groups, values, order=["WT", "KO", "DKO"],
                      show_mean=True)

    # log-scaled y axis
    pavlab_stripchart(groups, values, log_y="log2", ylabel="TPM")

    # save
    pavlab_stripchart(groups, values, filename="fig2b.svg", figsize=(4, 5))

    # embed in a multi-panel figure
    fig, axes = plt.subplots(1, 2, figsize=(10, 5))
    pavlab_stripchart(g1, v1, ax=axes[0])
    pavlab_stripchart(g2, v2, ax=axes[1])
"""

from __future__ import annotations

import warnings
from typing import Any, Union

import matplotlib.pyplot as plt
import numpy as np

try:
    from .pavlab_scatter import (
        _pick_log_kind,
        _apply_log,
        _axis_label,
        _font_sizes,
        _maybe_warn_log,
        _resolve_color,
        _BLACK,
        _N_ALPHA,
    )
except ImportError:
    from pavlab_scatter import (  # script / sys.path style
        _pick_log_kind,
        _apply_log,
        _axis_label,
        _font_sizes,
        _maybe_warn_log,
        _resolve_color,
        _BLACK,
        _N_ALPHA,
    )

# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------

_MEAN_COLOR = "#9ca3af"   # gray-400
_N_DENSITY  = 10_000      # alpha floor threshold (no hexbin mode in stripchart)

LogScale = Union[bool, str]   # False | True | "log2" | "log10"


# ---------------------------------------------------------------------------
# Main function
# ---------------------------------------------------------------------------

def pavlab_stripchart(
    x,
    y,
    color=None,
    color_label: str | None = None,
    jitter: float = 0.2,
    show_mean: bool = False,
    show_median: bool = False,
    order: list[str] | None = None,
    label_size: str | float = "med",
    xlabel: str | None = None,
    ylabel: str | None = None,
    title: str | None = None,
    log_y: LogScale = False,
    pseudocount: float = 1.0,
    origin_zero: bool = True,
    ylim: tuple | None = None,
    point_size: float | None = None,
    filename=None,
    figsize: tuple | None = None,
    ax=None,
) -> Any:
    """Pavlab-style strip chart (jittered categorical scatter).

    Parameters
    ----------
    x : array-like of str
        Group labels, one per observation.
    y : array-like of float
        Continuous values, one per observation.
    color : None | str | array-like, optional
        None → all black. A single matplotlib color string → uniform color.
        An array of strings → categorical colors (lab accent palette + legend).
        An array of floats → viridis colormap + colorbar.
    color_label : str, optional
        Colorbar label for continuous color.
    jitter : float
        Half-width of uniform horizontal jitter applied within each group
        column. Default 0.2. Set 0 to disable.
    show_mean : bool
        If True, draw the grand mean of all y values as a full-width dashed
        horizontal line (gray-400, linewidth=1, zorder below points).
    show_median : bool
        If True, draw the grand median as a full-width dotted horizontal line
        (same color and linewidth as mean line, zorder below points).
    order : list of str, optional
        Explicit group display order. Groups absent from order are silently
        dropped. Groups listed in order with no data are skipped with a warning.
    label_size : "small" | "med" | "large" | float
        Axis-label font size preset.
    xlabel, ylabel : str, optional
        Axis labels. Log-scale suffix appended automatically when a transform
        is applied.
    title : str, optional
        Figure title. Left-aligned, normal weight.
    log_y : False | True | "log2" | "log10"
        Log-transform the y axis. True auto-selects: log10 when max > 1 000
        (counts), log2 otherwise. pseudocount is added before transforming.
        A warning is issued when data spans > 100× without a transform.
    pseudocount : float
        Added to y values before log transform. Default 1.
    origin_zero : bool
        If True (default) and all post-transform y values ≥ 0, pin y-axis
        lower bound to 0. Explicit ylim= takes precedence.
    ylim : (lo, hi) | None
        Explicit y-axis limits.
    point_size : float | None
        Override automatic marker size. Default: s=20 for N ≤ 500, s=10
        otherwise (counted across all groups).
    filename : str | Path | None
        Save to file; format inferred from extension. Ignored when ax= is
        supplied.
    figsize : (w, h) | None
        Figure size in inches. Default (5.0, 5.0). Ignored when ax= is
        supplied.
    ax : matplotlib Axes | None
        Draw into an existing Axes.

    Returns
    -------
    matplotlib Axes
    """
    # ---- coerce inputs --------------------------------------------------
    x_raw = np.asarray(x, dtype=object).ravel()
    y_arr = np.asarray(y, dtype=float).ravel()
    if len(x_raw) != len(y_arr):
        raise ValueError(
            f"x and y must have equal length ({len(x_raw)} vs {len(y_arr)})"
        )

    color_raw = np.asarray(color) if color is not None else None
    if color_raw is not None and color_raw.ndim == 1 and len(color_raw) != len(x_raw):
        raise ValueError(
            f"color array length ({len(color_raw)}) must match x/y ({len(x_raw)})"
        )

    # ---- drop non-finite y ----------------------------------------------
    finite_mask = np.isfinite(y_arr)
    if not finite_mask.all():
        x_raw = x_raw[finite_mask]
        y_arr = y_arr[finite_mask]
        if color_raw is not None and color_raw.ndim == 1 and len(color_raw) == len(finite_mask):
            color_raw = color_raw[finite_mask]

    if len(y_arr) == 0:
        warnings.warn("pavlab_stripchart: no finite data points to plot.", stacklevel=2)

    # ---- resolve group order --------------------------------------------
    seen_groups: list[str] = list(dict.fromkeys(x_raw.tolist()))

    if order is not None:
        unique_groups = []
        for g in order:
            if g not in seen_groups:
                warnings.warn(
                    f"pavlab_stripchart: group {g!r} listed in order has no data; "
                    "skipping.",
                    stacklevel=2,
                )
            else:
                unique_groups.append(g)
        group_set = set(unique_groups)
        keep = np.array([g in group_set for g in x_raw.tolist()], dtype=bool)
        x_raw = x_raw[keep]
        y_arr = y_arr[keep]
        if color_raw is not None and color_raw.ndim == 1:
            color_raw = color_raw[keep]
    else:
        unique_groups = seen_groups

    n = len(y_arr)

    # ---- suggest / apply log on y --------------------------------------
    if log_y is False:
        _maybe_warn_log(y_arr, "y")

    y_kind = _pick_log_kind(log_y, y_arr, "y")
    if y_kind:
        neg = int(np.sum(np.isfinite(y_arr) & (y_arr + pseudocount <= 0)))
        if neg:
            warnings.warn(
                f"pavlab_stripchart: {neg} y value(s) will produce -inf after "
                f"log transform (pseudocount={pseudocount} insufficient).",
                stacklevel=2,
            )
        y_arr = _apply_log(y_arr, y_kind, pseudocount)

    # ---- build integer x positions + jitter ----------------------------
    group_index = {g: i for i, g in enumerate(unique_groups)}
    x_int = np.array([group_index[g] for g in x_raw.tolist()], dtype=float)
    if jitter > 0:
        x_int = x_int + np.random.uniform(-jitter, jitter, size=n)

    # ---- colour resolution ---------------------------------------------
    # Single color string must bypass _resolve_color, which treats a 0-d
    # string array as a 1-character iterable and produces garbage categories.
    if isinstance(color, str):
        c_arg, cmap_arg, norm_arg, legend_handles, is_single = color, None, None, None, True
    else:
        color_arg = color_raw if (color_raw is not None and color_raw.ndim == 1) else color
        c_arg, cmap_arg, norm_arg, legend_handles, is_single = _resolve_color(color_arg, n)

    # ---- marker size + alpha -------------------------------------------
    ps = float(point_size) if point_size is not None else (
        20.0 if n <= _N_ALPHA else 10.0
    )
    if n <= _N_ALPHA:
        alpha = 1.0
    else:
        frac = min(1.0, (n - _N_ALPHA) / max(1, _N_DENSITY - _N_ALPHA))
        alpha = max(0.15, 0.6 - frac * 0.45)

    # ---- figure setup --------------------------------------------------
    own_fig = ax is None
    if own_fig:
        fig, ax = plt.subplots(figsize=figsize or (5.0, 5.0))
    else:
        fig = ax.figure

    ax.set_facecolor("white")
    ax.grid(False)
    ax.spines["top"].set_visible(False)
    ax.spines["right"].set_visible(False)

    label_pt, tick_pt = _font_sizes(label_size)

    # ---- mean / median reference lines (below points) ------------------
    if (show_mean or show_median) and n > 0:
        x_span_lo = -0.5
        x_span_hi = len(unique_groups) - 0.5

    if show_mean and n > 0:
        grand_mean = float(np.mean(y_arr[np.isfinite(y_arr)]))
        ax.hlines(
            grand_mean,
            x_span_lo,
            x_span_hi,
            colors=_MEAN_COLOR,
            linewidths=1,
            linestyles="dashed",
            zorder=1,
        )

    if show_median and n > 0:
        grand_median = float(np.median(y_arr[np.isfinite(y_arr)]))
        ax.hlines(
            grand_median,
            x_span_lo,
            x_span_hi,
            colors=_MEAN_COLOR,
            linewidths=1,
            linestyles="dotted",
            zorder=1,
        )

    # ---- draw points ---------------------------------------------------
    sc_kw: dict[str, Any] = {
        "s": ps,
        "alpha": alpha,
        "linewidths": 0,
        "marker": "o",
        "zorder": 2,
    }
    if is_single:
        sc_kw["color"] = c_arg
    else:
        sc_kw["c"] = c_arg
        if cmap_arg is not None:
            sc_kw["cmap"] = cmap_arg
        if norm_arg is not None:
            sc_kw["norm"] = norm_arg

    sc = ax.scatter(x_int, y_arr, **sc_kw)

    if cmap_arg is not None:
        cb = fig.colorbar(sc, ax=ax)
        if color_label:
            cb.set_label(color_label, fontsize=tick_pt)

    if legend_handles:
        ax.legend(handles=legend_handles, frameon=False, fontsize=tick_pt)

    # ---- x-axis ticks + labels -----------------------------------------
    ax.set_xticks(range(len(unique_groups)))
    ax.set_xticklabels(unique_groups)
    ax.set_xlim(-0.5, len(unique_groups) - 0.5)

    # ---- y-axis limits -------------------------------------------------
    if ylim is not None:
        ax.set_ylim(ylim)
    else:
        fy = y_arr[np.isfinite(y_arr)]
        if origin_zero and fy.size > 0 and float(fy.min()) >= 0:
            ax.set_ylim(bottom=0)

    # ---- axis labels + title -------------------------------------------
    xl = xlabel or ""
    yl = _axis_label(ylabel, y_kind)
    if xl:
        ax.set_xlabel(xl, fontsize=label_pt)
    if yl:
        ax.set_ylabel(yl, fontsize=label_pt)
    if title:
        ax.set_title(title, fontsize=label_pt, loc="left", fontweight="normal")

    ax.tick_params(axis="both", labelsize=tick_pt)

    # ---- output --------------------------------------------------------
    if own_fig:
        fig.tight_layout()
        if filename is not None:
            fig.savefig(filename, dpi=200)
            plt.close(fig)

    return ax

File: skills/plotting/R/pavlab_stripchart.R
# pavlab_stripchart: opinionated lab-style strip chart (ggplot2 backend).
# Mirrors the design conventions of pavlab_scatter for categorical x / continuous y:
#   - Jittered points, no fill, shape 16, no stroke
#   - N-adaptive alpha and point size
#   - log_y: FALSE/TRUE/"log2"/"log10"; same semantics as pavlab_scatter
#   - color: NULL=black; single string=uniform; char/factor vec=categorical+legend;
#            numeric vec=viridis colorbar
#   - Optional grand-mean and grand-median reference lines (behind points)
#   - origin_zero: pins non-negative y lower bound to 0
#
# Helpers reused from pavlab_scatter.R (which sources palettes.R):
#   .pick_log_kind  .apply_log  .axis_label  .font_sizes  .maybe_warn_log
#   .color_mode  .N_ALPHA  .ACCENT_COLORS  .LABEL_SIZES
#
# Usage:
#   source("R/palettes.R")
#   source("R/pavlab_scatter.R")
#   source("R/pavlab_stripchart.R")
#   pavlab_stripchart(groups, values)
#   pavlab_stripchart(groups, values, log_y="log2", ylabel="Expression", show_mean=TRUE)
#   pavlab_stripchart(groups, values, color=group_vec, order=c("ctrl","treat"))
#   p <- pavlab_stripchart(groups, values, filename="fig2a.svg")
#   p + ggplot2::geom_boxplot(alpha=0)   # extend with ggplot2 layers

# ---- Main function -----------------------------------------------------------

#' Pavlab-style strip chart (ggplot2).
#'
#' Categorical x-axis, continuous y-axis; points jittered horizontally.
#' N-adaptive point size and alpha. Returns a ggplot2 object that can be
#' extended with `+` layers.
#'
#' @param x Character or factor vector of group labels (length n).
#' @param y Numeric vector of values (length n). Non-finite values are silently
#'   dropped; aligned `color` vector is masked to the same rows.
#' @param color NULL (all black), a single color string (uniform), a
#'   character / factor vector of length n (categorical — lab accent palette +
#'   legend), or a numeric vector of length n (viridis colorbar).
#' @param color_label Colorbar title for numeric `color`. Default NULL (empty).
#' @param jitter Horizontal jitter width passed to `position_jitter`. Default 0.2.
#' @param show_mean If TRUE, draw a dashed horizontal reference line at the
#'   grand mean of all (post-transform) y values. Line renders behind points.
#'   Default FALSE.
#' @param show_median If TRUE, draw a dotted horizontal reference line at the
#'   grand median. Default FALSE.
#' @param order Optional character vector of group levels controlling x-axis
#'   order. Unrecognised levels in `x` are silently appended in appearance order.
#' @param log_y FALSE | TRUE | "log2" | "log10". TRUE auto-selects: log10 when
#'   max > 1 000, log2 otherwise. pseudocount added before transform so zeros
#'   survive. Warns when data spans > 100× and no transform is requested.
#' @param pseudocount Added to y before log transform. Default 1.
#' @param origin_zero If TRUE (default) and all transformed y >= 0, expand the
#'   y lower bound to include 0. Explicit ylim takes precedence.
#' @param ylim Numeric length-2 vector for explicit y limits. Takes precedence
#'   over origin_zero.
#' @param label_size "small", "med", "large", or a numeric point size for axis
#'   labels (tick labels scale proportionally). Default "med" (14 pt / 11 pt).
#' @param xlabel,ylabel Axis labels. Log-scale suffix ("(log₂)" etc.) appended
#'   automatically on the y axis when a transform is applied.
#' @param title Plot title. Left-aligned, normal weight.
#' @param point_size Override auto marker size (ggplot2 `size=` in mm).
#'   Default: 2.0 for N <= 500, 1.5 otherwise.
#' @param filename Save path. Format inferred from extension via ggsave.
#' @param figsize Numeric length-2 vector c(width, height) in inches.
#'   Default c(5.5, 5.0).
#' @return ggplot2 object (invisibly).
#' @export
pavlab_stripchart <- function(
  x, y,
  color       = NULL,
  color_label = NULL,
  jitter      = 0.2,
  show_mean   = FALSE,
  show_median = FALSE,
  order       = NULL,
  log_y       = FALSE,
  pseudocount = 1.0,
  origin_zero = TRUE,
  ylim        = NULL,
  label_size  = "med",
  xlabel      = NULL,
  ylabel      = NULL,
  title       = NULL,
  point_size  = NULL,
  filename    = NULL,
  figsize     = NULL
) {
  if (!requireNamespace("ggplot2", quietly = TRUE))
    stop("pavlab_stripchart: ggplot2 is required. Install with install.packages('ggplot2').",
         call. = FALSE)

  x <- as.character(x)
  y <- as.numeric(y)
  if (length(x) != length(y))
    stop(sprintf("x and y must have equal length (%d vs %d)", length(x), length(y)),
         call. = FALSE)

  # Drop non-finite y; align color vector to the same mask.
  has_vec_color <- !is.null(color) && length(color) == length(y)
  mask <- is.finite(y)
  if (has_vec_color) color <- color[mask]
  x <- x[mask]
  y <- y[mask]

  n <- length(y)
  if (n == 0)
    warning("pavlab_stripchart: no finite data points to plot.", call. = FALSE)

  # Suggest log scale when not transforming.
  if (identical(log_y, FALSE)) .maybe_warn_log(y, "y")

  # Apply log transform.
  y_kind <- .pick_log_kind(log_y, y, "y")
  if (!is.null(y_kind)) {
    neg <- sum(is.finite(y) & (y + pseudocount) <= 0)
    if (neg > 0)
      warning(sprintf(
        "pavlab_stripchart: %d y value(s) will produce -Inf after log transform (pseudocount=%g insufficient).",
        neg, pseudocount
      ), call. = FALSE)
    y <- .apply_log(y, y_kind, pseudocount)
  }

  # Build factor levels from order, then append any remaining values in
  # appearance order so nothing gets silently dropped.
  seen <- unique(x)
  if (!is.null(order)) {
    extra  <- seen[!seen %in% order]
    levels <- c(order, extra)
  } else {
    levels <- seen
  }
  x_fac <- factor(x, levels = levels)

  # Font sizes.
  fs       <- .font_sizes(label_size)
  label_pt <- fs[1]
  tick_pt  <- fs[2]

  # Y-axis label with optional log suffix.
  yl <- .axis_label(ylabel, y_kind)

  # Point size + alpha.
  ps <- if (!is.null(point_size)) as.numeric(point_size) else
          if (n <= .N_ALPHA) 2.0 else 1.5

  alpha_val <- if (n <= .N_ALPHA) {
    1.0
  } else {
    frac <- min(1.0, (n - .N_ALPHA) / max(1L, 10000L - .N_ALPHA))
    max(0.15, 0.6 - frac * 0.45)
  }

  # Build data frame.
  df <- data.frame(x_grp = x_fac, y_val = y, stringsAsFactors = FALSE)

  # ---- Base ggplot + reference lines (drawn first, behind points) ----------

  cmode <- .color_mode(color, n)

  # Construct the base aesthetic; color column added per mode below.
  if (cmode %in% c("uniform_black", "uniform_string")) {
    p <- ggplot2::ggplot(df, ggplot2::aes(x = x_grp, y = y_val))
  } else if (cmode == "categorical") {
    df$color_col <- as.character(color)
    p <- ggplot2::ggplot(df, ggplot2::aes(x = x_grp, y = y_val, color = color_col))
  } else {
    df$color_col <- as.numeric(color)
    p <- ggplot2::ggplot(df, ggplot2::aes(x = x_grp, y = y_val, color = color_col))
  }

  # Reference lines go before geom_jitter so they render behind the points.
  if (show_mean) {
    grand_mean <- mean(df$y_val, na.rm = TRUE)
    p <- p + ggplot2::geom_hline(
      yintercept = grand_mean,
      color      = "#9ca3af",
      linetype   = "dashed",
      linewidth  = 0.5
    )
  }
  if (show_median) {
    grand_median <- stats::median(df$y_val, na.rm = TRUE)
    p <- p + ggplot2::geom_hline(
      yintercept = grand_median,
      color      = "#9ca3af",
      linetype   = "dotted",
      linewidth  = 0.5
    )
  }

  # ---- Points --------------------------------------------------------------

  jitter_pos <- ggplot2::position_jitter(width = jitter, height = 0)

  if (cmode == "uniform_black") {
    p <- p + ggplot2::geom_jitter(
      color    = "#000000",
      alpha    = alpha_val,
      size     = ps,
      shape    = 16,
      stroke   = 0,
      position = jitter_pos
    )

  } else if (cmode == "uniform_string") {
    p <- p + ggplot2::geom_jitter(
      color    = as.character(color),
      alpha    = alpha_val,
      size     = ps,
      shape    = 16,
      stroke   = 0,
      position = jitter_pos
    )

  } else if (cmode == "categorical") {
    cats     <- unique(as.character(color))
    cat_idx  <- ((seq_along(cats) - 1L) %% length(.ACCENT_COLORS)) + 1L
    cat_cols <- setNames(.ACCENT_COLORS[cat_idx], cats)
    p <- p +
      ggplot2::geom_jitter(
        alpha    = alpha_val,
        size     = ps,
        shape    = 16,
        stroke   = 0,
        position = jitter_pos
      ) +
      ggplot2::scale_color_manual(values = cat_cols, name = NULL) +
      ggplot2::guides(color = ggplot2::guide_legend(
        override.aes = list(size = 3, alpha = 1)
      ))

  } else {
    # numeric → viridis colorbar
    clabel <- if (!is.null(color_label)) color_label else ""
    p <- p +
      ggplot2::geom_jitter(
        alpha    = alpha_val,
        size     = ps,
        shape    = 16,
        stroke   = 0,
        position = jitter_pos
      ) +
      ggplot2::scale_color_viridis_c(name = clabel)
  }

  # ---- Theme ---------------------------------------------------------------

  p <- p +
    ggplot2::theme_classic(base_size = tick_pt) +
    ggplot2::theme(
      axis.title        = ggplot2::element_text(size = label_pt, color = "#1f2937"),
      axis.text         = ggplot2::element_text(size = tick_pt,  color = "#6b7280"),
      axis.ticks        = ggplot2::element_line(color = "#6b7280"),
      plot.title        = ggplot2::element_text(size = label_pt, color = "#1f2937",
                                                hjust = 0, face = "plain"),
      legend.text       = ggplot2::element_text(size = tick_pt),
      legend.title      = ggplot2::element_text(size = tick_pt),
      legend.background = ggplot2::element_blank(),
      legend.key        = ggplot2::element_blank(),
      plot.margin       = ggplot2::margin(3, 3, 3, 3, "mm")
    )

  # ---- Labels --------------------------------------------------------------

  if (!is.null(xlabel) && nchar(xlabel) > 0) p <- p + ggplot2::xlab(xlabel)
  if (nchar(yl) > 0) p <- p + ggplot2::ylab(yl)
  if (!is.null(title)) p <- p + ggplot2::ggtitle(title)

  # ---- Y-axis limits -------------------------------------------------------

  if (origin_zero) {
    fy <- df$y_val[is.finite(df$y_val)]
    if (length(fy) > 0 && min(fy) >= 0) p <- p + ggplot2::expand_limits(y = 0)
  }
  if (!is.null(ylim)) {
    p <- p + ggplot2::coord_cartesian(ylim = ylim)
  }

  # ---- Save ----------------------------------------------------------------

  if (!is.null(filename)) {
    w <- if (!is.null(figsize)) figsize[1] else 5.5
    h <- if (!is.null(figsize)) figsize[2] else 5.0
    ggplot2::ggsave(filename, plot = p, width = w, height = h, dpi = 150)
  }

  invisible(p)
}
